spyder/app/mainwindow.py

```diff
@@ -250,6 +250,157 @@ class MainWindow(QMainWindow):
+        # TODO: Handle by class or by name
+        for plugin_name in self._PLUGINS:
+            try:
+                if plugin_class.NAME == plugin_name:
+                    return self._PLUGINS[plugin_name]
+            except AttributeError:
+                if plugin_class.CONF_SECTION == plugin_name:
+                    return self._PLUGINS[plugin_name]
+        else:
+            raise Exception('Plugin not found!')
+
+    def show_status_message(self, message, timeout):
+        """
+        Show a status message in Spyder Main Window.
+        """
+        status_bar = self.statusBar()
+        if status_bar.isVisible():
+            status_bar.showMessage(message, timeout)
+
+    def show_plugin_compatibility_message(self, message):
+        """
+        Show a compatibility message.
+        """
+        messageBox = QMessageBox(self)
+        messageBox.setWindowModality(Qt.NonModal)
+        messageBox.setAttribute(Qt.WA_DeleteOnClose)
+        messageBox.setWindowTitle(_('Compatibility Check'))
+        messageBox.setText(message)
+        messageBox.setStandardButtons(QMessageBox.Ok)
+        messageBox.show()
+
+    def register_plugin(self, plugin):
+        """
+        Register a plugin in Spyder Main Window.
+        """
+        self.set_splash(_("Loading {}...".format(plugin.get_name())))
+        logger.info("Loading {}...".format(plugin.NAME))
+
+        # Check plugin compatibility
+        is_compatible, message = plugin.check_compatibility()
+        plugin._is_compatible = is_compatible
+        if not is_compatible:
+            self.show_compatibility_message(message)
+            return None
+
+        # Signals
+        plugin.sig_status_message_requested.connect(self.show_status_message)
+        plugin.sig_quit_requested.connect(self.close)
+
+        # Register plugin
+        plugin._register()
+        plugin.register()
+
+        # Add dockwidget
+        logger.info("Adding dockwidget for {}...".format(plugin.NAME))
+        self.add_dockwidget(plugin)
+
+        # Update margins
+        margin = 0
+        if CONF.get('main', 'use_custom_margin'):
+            margin = CONF.get('main', 'custom_margin')
+        plugin.update_margins(margin)
+
+        # First time plugin starts
+        if plugin.get_option('first_time', True):
+            try:
+                logger.info("Tabify {} dockwidget...".format(plugin.NAME))
+                next_to_plugin = self.get_plugin(plugin.TABIFY)
+                self.tabify_plugins(plugin, next_to_plugin)
+                # Do we really need this?
+                # plugin.on_first_registration()
+            except NotImplementedError:
+                return
+            plugin.set_option('first_time', False)
+
+        for action_name, action in plugin.get_actions().items():
+            context = (getattr(action, 'shortcut_context', plugin.NAME)
+                       or plugin.NAME)
+            self.register_shortcut(action, context, action_name)
+

```

Comment:
Here is where "exposed actions" are registered as shortcuts.

I need to do some cleanup but the idea is the following.

- Actions should be created once (not destroyed and recycled on popup of menus etc,...)
- To create an action the user must user `self.create_action` by using the SpyderWidgetMixin.
- A PluginMainWidget will collect any actions defined in subwidgets (if defined) and expose them in the `get_actions` method. This could be even automated so that we find all children widgets and check if they define these actions. By doing it like this, users do not need to worry about exposing these actions manually but the API does it for them
- Any action created this way is now exposed as a possible shortcut automatically without manual registration. If an option is found in the config then it is assigned otherwise is left with an empty shortcut. By doing it like this the `name` of actions is now also used for the `name`of the shortcut and by doing that we already have the translation of the shortcut via the action, plus we now never need to worry about these shortcuts.

Is it more clear now?

Reply:
Yes, thank you. If you can write part of this in the doc string of `create_action` and `get_actions`, that would be grand.

Reply:
Will do!